### PR TITLE
feat: use actuator/metrics for data points

### PR DESCRIPTION
**Pull request by clemens-tolboom**
### Current

image

### Previous

image

### Other Actuator

This makes discussion easier as the links are available.

image

We can use de standard items and build custom

- [x] use standard
- [x] Through new issue ~~make custom rock items available by their API like `rserve.processes.current` already there~~
- [x] make UI table
- [ ] Add tests
- [ ] cleanup UI
- [ ] Align header same as Projects 
- [ ] Add search to filter on
  - image
- [x] Details clutters the table feel
  - image
  - [ ] Move Details in column as popout?
- [ ] odd/even on path difference .. detail above has 2 items on same path
- [ ] Bytes are rounded up to 2 digits precision B,KB,MB,GM,TB,EB
  - [ ] is similar nice to have for times ie uptime in days?
image

---

# Actuator
### By names

```bash
curl --silent --user admin:admin http://localhost:8080/actuator/metrics | json_pp
{
   "names" : [
      "application.ready.time",
      "application.started.time",
      "disk.free",
      "disk.total",
      "http.server.requests",
      "http.server.requests.active",
      "jvm.buffer.count",
      "jvm.buffer.memory.used",
...
```
#### fetch by name
```bash
# file: actuator-metrics
json=$(curl --silent --user admin:admin http://localhost:8080/actuator/metrics)

names=$(echo "$json" | jq -r '.names[]')
```bash
for name in $names; do
   echo "Processing $name"
   # Add your processing code here
done
```
```

```
### Groups

```bash
curl --silent --user admin:admin http://localhost:8080/actuator/metrics|
json_pp | grep '\.' | cut -d . -f 1 | uniq -c

   2       "application
   2       "disk
   2       "http
  22       "jvm
   1       "logback
   5       "process
   1       "rserve
  42       "spring
   3       "system
   6       "tomcat
```

### disk.free

```bash
curl --silent --user admin:admin http://localhost:8080/actuator/metrics/disk.free | json_pp

{
   "availableTags" : [
      {
         "tag" : "path",
         "values" : [
            "/Users/clemens/Documents/GitHub/molgenis-service-armadillo/."
         ]
      }
   ],
   "baseUnit" : "bytes",
   "description" : "Usable space for path",
   "measurements" : [
      {
         "statistic" : "VALUE",
         "value" : 269983137792
      }
   ],
   "name" : "disk.free"
}
```

### CODE REVIEW

**Comment by clemens-tolboom** [comment]
Weird error while testing on `rserve.processes.current` ... is that an actuator item?

```
15:45:45.103 [http-nio-8080-exec-4|] WARN  o.m.r.RServerConnectionFactory - Container for 'http://localhost:6312'  is down
15:45:45.118 [http-nio-8080-exec-4|] WARN  i.m.c.i.internal.DefaultGauge - Failed to apply the value function for the gauge 'rserve.processes.current'. Note that subsequent logs will be logged at debug level.
org.molgenis.r.exceptions.ConnectionCreationFailedException: org.molgenis.r.rock.RockServerException: Failure when opening a Rock R session
        at org.molgenis.r.rock.RockConnectionFactory.tryCreateConnection(RockConnectionFactory.java:39)
        at org.molgenis.r.RServerConnectionFactory.tryCreateConnection(RServerConnectionFactory.java:66)
        at org.molgenis.armadillo.info.RProcessEndpoint.connect(RProcessEndpoint.java:58)
...
```

File: armadillo/src/main/java/org/molgenis/armadillo/info/FileMetrics.java
Comment (marikaris) [inline armadillo/src/main/java/org/molgenis/armadillo/info/FileMetrics.java:11]: Could you add a test for this?

File: ui/src/api/api.ts
Comment (marikaris) [inline ui/src/api/api.ts:116]: Could we fix this before merge?

File: ui/src/api/api.ts
Comment (marikaris) [inline ui/src/api/api.ts:140]: Why do we do we alter the output here and don't we create the right output in the backend? (just curious, you probably have a good reason)

File: ui/src/api/api.ts
Comment (marikaris) [inline ui/src/api/api.ts:157]: we want to be able to see errors in the ui, so they need to be passed through the api call

File: ui/src/api/api.ts
Comment (marikaris) [inline ui/src/api/api.ts:163]: return the get, not the await, so we can show loading spinner

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:19]: remove console.log

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:13]: Not entirely sure what you're trying to do here. Feels duplicate somehow. You're trying to add the key as a value to the item? Why do we need this? Can't we solve this when we're processing the data?

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:32]: Please use `const` and `let` instead of `var`

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:60]: please write a test for this

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:64]: No TODO's in prod code.

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:68]: what does this `FIELD_DISPLAY` thing do?

File: ui/src/components/Actuator.vue
Comment (marikaris) [inline ui/src/components/Actuator.vue:127]: please add key

File: ui/src/components/ActuatorItem.vue
Comment (marikaris) [inline ui/src/components/ActuatorItem.vue:17]: this seems like a helper function to me, please move to helper/utils and add test

File: ui/src/components/ActuatorItem.vue
Comment (marikaris) [inline ui/src/components/ActuatorItem.vue:30]: Please remove uncommented code

File: ui/src/components/ActuatorItem.vue
Comment (marikaris) [inline ui/src/components/ActuatorItem.vue:52]: please add key

File: ui/src/components/ActuatorItem.vue
Comment (marikaris) [inline ui/src/components/ActuatorItem.vue:66]: Ah! This is why you use that display thingy, right? I would set this up a bit differently, like we did throughout the rest of the application. I would compute the filtered items or create a method, then only show those and in case there is no searchword, then filtered items should return the full list. If you do this, you don't have to worry about resetting things when the search is emptied. 
Example: https://github.com/molgenis/molgenis-service-armadillo/blob/23272d29bb66cc343e058587ad9dc5945897c99b/ui/src/views/Projects.vue#L310
I also changed it on the branch I created for this, so you could also look at that.

File: ui/src/views/Insight.vue
Comment (marikaris) [inline ui/src/views/Insight.vue:3]: I don't think you should use tabs here. Tabs in tabs are generally confusing, but I think we already agreed upon a solution :)

File: ui/src/views/Insight.vue
Comment (marikaris) [inline ui/src/views/Insight.vue:45]: why is this there? don't think we're using the red border thingy, so we don't need the style?

File: ui/src/views/Users.vue
Comment (marikaris) [inline ui/src/views/Users.vue:128]: please remove whitespace

**Review by marikaris — CHANGES_REQUESTED** [review]
As discussed. For now I'm not going to comment on the composition/options api. We will the refactor for that later. I do think we need tests and the typing should be fixed. Do you have the following plugins in your vsCode:
`TypeScript Vue Plugin` and `Vue Language Features (Volar)` ? I think it's the latter. I get error like these: 
Screenshot 2024-02-26 at 10 43 25
 :
Screenshot 2024-02-26 at 10 42 59
Screenshot 2024-02-26 at 10 43 07
Screenshot 2024-02-26 at 10 43 17

File: ui/src/api/api.ts
Comment (clemens-tolboom) [reply ui/src/api/api.ts:152]: That was bad code indeed and is fixed now

File: ui/src/api/api.ts
Comment (clemens-tolboom) [reply ui/src/api/api.ts:157]: Fixed

File: ui/src/api/api.ts
Comment (clemens-tolboom) [reply ui/src/api/api.ts:179]: Fixed

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:1]: Done

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue]: Done

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:39]: Done

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue]: I disagree as a TODO can be needed later on in the project.

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:155]: Is for next phase

File: ui/src/components/ActuatorItem.vue
Comment (clemens-tolboom) [reply ui/src/components/ActuatorItem.vue:45]: It's a one of AFAIK so local helper

File: ui/src/components/ActuatorItem.vue
Comment (clemens-tolboom) [reply ui/src/components/ActuatorItem.vue:58]: This is structure docs for discussing the template so keep it for now

**Review by mswertz — APPROVED** [review]
approve but please check if you can add smoketest for the '/actuator' and '/metrics'
I expect you would like to copy the test for the UserController

File: armadillo/src/main/java/org/molgenis/armadillo/info/FileMetrics.java
Comment (clemens-tolboom) [reply armadillo/src/main/java/org/molgenis/armadillo/info/FileMetrics.java:11]: will do in next iteration.

File: ui/src/api/api.ts
Comment (clemens-tolboom) [reply ui/src/api/api.ts:128]: Done

File: ui/src/api/api.ts
Comment (clemens-tolboom) [reply ui/src/api/api.ts]: Done

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue]: Gone

File: ui/src/components/ActuatorItem.vue
Comment (clemens-tolboom) [reply ui/src/components/ActuatorItem.vue]: Done

File: ui/src/components/ActuatorItem.vue
Comment (clemens-tolboom) [reply ui/src/components/ActuatorItem.vue]: We can check later in merged code

File: ui/src/views/Users.vue
Comment (clemens-tolboom) [reply ui/src/views/Users.vue:128]: Removed

Comment (clemens-tolboom) [reply ui/src/views/Users.vue:128]: I keep it as it separates libs from components

File: ui/src/views/Insight.vue
Comment (clemens-tolboom) [reply ui/src/views/Insight.vue]: Removed

File: ui/src/views/Insight.vue
Comment (clemens-tolboom) [reply ui/src/views/Insight.vue]: Will fix after merge

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue]: done

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:87]: Adding the key into the item

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:90]: Code is different but I remember `api` failure must end in the UI not processed in api itself. I partly fixed that.

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:4]: I'm not sure how to add error within component ... will discuss after merge

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:90]: Test is #701 which fails for some weird reasons ... will make separate PR

File: ui/src/components/Actuator.vue
Comment (clemens-tolboom) [reply ui/src/components/Actuator.vue:4]: Is followup issue #702

**Pull request merged**
